src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            linkFirst(value);
+        } else {
+            linkLast(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (isEmpty()) {
+            linkFirst(value);
+        } else if (index == size()) {
+            linkLast(value);
+        } else if (index == 0) {
+            Node<T> current = first;
+            Node<T> newNode = new Node<>(null, value, current);
+            first = newNode;
+            current.prev = newNode;
+        } else {
+            Node<T> current = getNode(index);
+            Node<T> newNode = new Node<>(current.prev, value, current);
+            current.prev.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForGet(index);
+        Node<T> current = first;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current != null ? current.value : null;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexForGet(index);
+        Node<T> current = getNode(index);
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexForGet(index);
+        Node<T> current = getNode(index);
+        T removeValue = current.value;
+        if (current == first && size > 1) {
+            current.next.prev = null;
+            first = current.next;
+        } else if (current == last && size > 1) {
+            current.prev.next = null;
+            last = current.prev;
+        } else if (size > 1) {
+            current.next.prev = current.prev;
+            current.prev.next = current.next;
+        }
+        size--;
+        return removeValue;
```

Comment:
I cannot see any actions done if current size is 1 and passed index is 0, which is valid case